### PR TITLE
Vehicle create

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Okay I promise not to abandon this one for two years...

Continues on from the other vehicle PR.

### CODE REVIEW

File: src/openloco/things/vehicle.h [hunk 1]
```diff
@@ -185,5 +216,49 @@ namespace openloco
...
+        uint8_t pad_5A[0x5E - 0x5A];
+        uint8_t var_5E;
+        uint8_t var_5F;
+        uint8_t var_60;
+        uint8_t var_61;
+        uint32_t var_62;
+        uint16_t reliability; // 0x66
+        uint16_t var_68;
+    };
+    static_assert(offsetof(vehicle_bogie, var_68) == 0x68);
```
Comment (AaronVanGeffen) [inline src/openloco/things/vehicle.h:262]: Travis CI is failing on these. Perhaps use something like the following instead:
```suggestion
    static_assert(sizeof(vehicle_bogie) == 0x69);
```

File: src/openloco/things/CreateVehicle.cpp [hunk 2]
```diff
@@ -0,0 +1,990 @@
...
+using namespace openloco::objectmgr;
+
+namespace openloco::things::vehicle
+{
+    static loco_global<company_id_t, 0x009C68EB> _updating_company_id;
+    static loco_global<uint16_t, 0x009C68E0> gameCommandMapX;
+    static loco_global<uint16_t, 0x009C68E2> gameCommandMapY;
+    static loco_global<uint16_t, 0x009C68E4> gameCommandMapZ;
+    static loco_global<string_id, 0x009C68E6> gGameCommandErrorText;
+    static loco_global<uint8_t, 0x009C68EA> gGameCommandExpenditureType; // multiplied by 4
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:29]:
```suggestion
    static loco_global<uint8_t, 0x009C68EA> gGameCommandExpenditureType; // premultiplied by 4
```

File: src/openloco/things/CreateVehicle.cpp [hunk 3]
```diff
@@ -0,0 +1,990 @@
...
+                thingmgr::freeThing(veh1);
+                auto veh6 = veh2->next_car();
+                thingmgr::freeThing(veh2);
+                thingmgr::freeThing(veh6);
+                return 0x80000000;
+            }
+        }
+        // 0x4AE733
+        auto vehObject = objectmgr::get<vehicle_object>(vehicleTypeId);
+        auto cost = (vehObject->cost_fact * currencyMultiplicationFactor[vehObject->cost_ind]) / 64;
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:896]:
```suggestion
        // TODO: use FixedPoint with 6 decimals for cost_ind
        auto cost = (vehObject->cost_fact * currencyMultiplicationFactor[vehObject->cost_ind]) / 64;
```

File: src/openloco/things/CreateVehicle.cpp [hunk 4]
```diff
@@ -0,0 +1,990 @@
...
+    }
+    // 0x004AE318
+    static std::optional<openloco::vehicle_head*> createBase(const TransportMode mode, const VehicleType type, const uint8_t trackType)
+    {
+        if (!thingmgr::checkNumFreeThings(4))
+        {
+            return {};
+        }
+
+        if (_525FB8 > 256000)
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:699]:
```suggestion
        if (_525FB8 >= 256000)
```

```asm
cmp     eax, 256000
jnb     loc_4AE5D9
```

File: src/openloco/things/CreateVehicle.cpp [hunk 5]
```diff
@@ -0,0 +1,990 @@
...
+            if (id == 0xFFFF)
+            {
+                return true;
+            }
+        }
+        gGameCommandErrorText = string_ids::too_many_vehicles;
+        return false;
+    }
+
+    static bool sub_431E6A(const company_id_t company, map::tile_element* const tile)
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:91]: Add arguments (al, esi) to comment

File: src/openloco/things/CreateVehicle.cpp [hunk 6]
```diff
@@ -0,0 +1,990 @@
...
+    template<typename T>
+    static T* createVehicleThing()
+    {
+        auto* const base = createVehicleThing();
+        base->base_type = thing_base_type::vehicle;
+        base->type = T::VehicleThingType;
+        return reinterpret_cast<T*>(base);
+    }
+
+    static void sub_4BA873(openloco::vehicle_bogie* const vehBogie)
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:195]: Too math intensive for me. Someone else validate please.

File: src/openloco/things/CreateVehicle.cpp [hunk 7]
```diff
@@ -0,0 +1,990 @@
...
+            auto& prng = gprng();
+            int32_t randVal = (prng.rand_next(65535) * reliabilityFactor / 2) / 256;
+            reliabilityFactor -= reliabilityFactor / 4;
+            reliabilityFactor += randVal;
+            vehBogie->var_68 = static_cast<uint16_t>(std::max(4, reliabilityFactor));
+        }
+    }
+
+    // 0x4AE8F1
+    static openloco::vehicle_bogie* createBogie(const thing_id_t head, const uint16_t vehicleTypeId, const vehicle_object& vehObject, const uint8_t bodyNumber, openloco::vehicle* const lastVeh, const ColourScheme colourScheme)
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:213]: Got until here from the top of the file

File: src/openloco/things/vehicle.cpp [hunk 8]
```diff
@@ -1546,3 +1547,165 @@ void openloco::vehicle_body::ship_wake_animation_update(uint8_t num, int8_t var_
...
+        return false;
+    }
+
+    if (newObject->type != vehicleType)
+    {
+        gGameCommandErrorText = string_ids::incompatible_vehicle;
+        return false;
+    }
+
+    {
```
[the remark this answers is not in the document]
Comment (duncanspumpkin) [reply src/openloco/things/vehicle.cpp:1672]: It was from an older version of the function before I broke it into more functions this allowed for using the same variable names again. I find it helps for splitting up the function.

File: src/openloco/things/vehicle.cpp [hunk 9]
```diff
@@ -1546,3 +1547,165 @@ void openloco::vehicle_body::ship_wake_animation_update(uint8_t num, int8_t var_
...
+        for (auto i = 0; i < newObject->num_compat; ++i)
+        {
+            if (newObject->compatible_vehicles[i] == sourceVehicleTypeId)
+            {
+                return true;
+            }
+        }
+    }
+
+    if (sourceObject->num_compat != 0)
```
[the remark this answers is not in the document]
Comment (duncanspumpkin) [reply src/openloco/things/vehicle.cpp:1582]: no will be removed.

File: src/openloco/things/CreateVehicle.cpp [hunk 10]
```diff
@@ -0,0 +1,990 @@
...
+    static bool checkMaxAIVehicles()
+    {
+        if (is_player_company(_updating_company_id))
+        {
+            return true;
+        }
+
+        const auto& companies = companymgr::companies();
+        auto totalAiVehicles = std::accumulate(companies.begin(), companies.end(), 0, [](int32_t& total, const auto& company) {
+            if (is_player_company(company.id()))
```
[the remark this answers is not in the document]
Comment (duncanspumpkin) [reply src/openloco/things/CreateVehicle.cpp:57]: It also misses adding the accumulator!

File: src/openloco/things/CreateVehicle.cpp [hunk 11]
```diff
@@ -0,0 +1,990 @@
...
+    static loco_global<uint8_t, 0x01136258> _backupZ;
+    static loco_global<uint16_t, 0x0113642A> _113642A; // used by build window and others
+    static loco_global<uint32_t[32], 0x00525E5E> currencyMultiplicationFactor;
+    static loco_global<uint8_t, 0x00525FC5> _525FC5;
+    static loco_global<uint32_t, 0x00525FB8> _525FB8;
+    static loco_global<thing_id_t[64 * 1000], 0x0096885C> _96885C;
+    static loco_global<uint8_t[256000], 0x00987C5C> _987C5C;
+
+    // 0x004B1D96
+    static bool checkMaxAIVehicles()
```
[the remark this answers is not in the document]
Comment (duncanspumpkin) [reply src/openloco/things/CreateVehicle.cpp:48]: I've made it aiIsBelowVehicleLimit as the player skips the whole check.

File: src/openloco/things/CreateVehicle.cpp (hunk 6 above)
Comment (duncanspumpkin) [reply src/openloco/things/CreateVehicle.cpp:195]: L214 looks incorrect should be `/ 65536`

File: src/openloco/things/CreateVehicle.cpp [hunk 12]
```diff
@@ -0,0 +1,1000 @@
...
+        if (newBogie == nullptr) // Can never happen
+        {
+            return nullptr;
+        }
+        newBogie->var_38 = 0;
+
+        int32_t reliability = vehObject.reliability * 256;
+        if (current_year() + 2 > vehObject.designed)
+        {
+            reliability -= reliability / 8;
```
Comment (duncanspumpkin) [inline src/openloco/things/CreateVehicle.cpp:273]: add comment. // Remove an eighth of reliability after 2 years past designed life

File: src/openloco/things/CreateVehicle.cpp [hunk 13]
```diff
@@ -0,0 +1,1000 @@
...
+            if (current_year() + 3 > vehObject.designed)
+            {
+                reliability -= reliability / 8;
```
Comment (duncanspumpkin) [inline src/openloco/things/CreateVehicle.cpp:276]: // Remove an eighth of reliability after 3 years past designed life

File: src/openloco/things/CreateVehicle.cpp [hunk 14]
```diff
@@ -0,0 +1,1000 @@
...
+            }
+        }
+        if (reliability != 0)
+        {
+            reliability += 255;
+        }
+        newBogie->reliability = reliability;
+        sub_4BA873(newBogie);
+
+        auto cost = (vehObject.cost_fact * currencyMultiplicationFactor[vehObject.cost_ind]) / 64;
```
Comment (duncanspumpkin) [inline src/openloco/things/CreateVehicle.cpp:286]: Add TODO's for all of these.

File: src/openloco/things/CreateVehicle.cpp [hunk 15]
```diff
@@ -0,0 +1,1000 @@
...
+        newBogie->var_62 = cost - cost / 8;
```
Comment (duncanspumpkin) [inline src/openloco/things/CreateVehicle.cpp:287]: Refund cost

File: src/openloco/things/CreateVehicle.cpp [hunk 16]
```diff
@@ -0,0 +1,1000 @@
...
+
+        if (bodyNumber == 0)
```
Comment (duncanspumpkin) [inline src/openloco/things/CreateVehicle.cpp:289]: Add comment about first bogie on first body has to store the secondary cargo.

File: src/openloco/things/CreateVehicle.cpp [hunk 17]
```diff
@@ -0,0 +1,1000 @@
...
+        //        }
+        //    }
+        //}
+        registers regs{};
+        regs.esi = reinterpret_cast<int32_t>(head);
+        call(0x004B7CC3, regs);
+    }
+
+    // 0x004AE86D
+    static bool createBody(openloco::vehicle_head* const head, const uint16_t vehicleTypeId)
```
Comment (duncanspumpkin) [inline src/openloco/things/CreateVehicle.cpp:456]: NextComponent instead of next car.

File: src/openloco/things/CreateVehicle.cpp [hunk 18]
```diff
@@ -0,0 +1,1000 @@
...
+        }
+        return newBogie;
+    }
+
+    // 0x004AEA9E
+    static openloco::vehicle_body* createBody(const thing_id_t head, const uint16_t vehicleTypeId, const vehicle_object& vehObject, const uint8_t bodyNumber, openloco::vehicle* const lastVeh, const ColourScheme colourScheme)
+    {
+        auto newBody = createVehicleThing<vehicle_body>();
+        // TODO: move this into the create function somehow
+        newBody->type = bodyNumber == 0 ? vehicle_thing_type::vehicle_body_end : vehicle_thing_type::vehicle_body_cont;
```
Comment (duncanspumpkin) [inline src/openloco/things/CreateVehicle.cpp:337]: Rename this to body start.

File: src/openloco/things/CreateVehicle.cpp [hunk 19]
```diff
@@ -0,0 +1,1000 @@
...
+            newBody->var_38 |= (1 << 3);
+        }
+
+        lastVeh->next_car_id = newBody->id; // same
+        return newBody;
+    }
+
+    static void sub_4B7CC3(openloco::vehicle_head* const head)
+    {
+        //if (veh0->mode == TransportMode::road)
```
Comment (duncanspumpkin) [inline src/openloco/things/CreateVehicle.cpp:419]: delete

File: src/openloco/things/CreateVehicle.cpp [hunk 20]
```diff
@@ -0,0 +1,1000 @@
...
+        //}
+        registers regs{};
+        regs.esi = reinterpret_cast<int32_t>(head);
+        call(0x004B7CC3, regs);
+    }
+
+    // 0x004AE86D
+    static bool createBody(openloco::vehicle_head* const head, const uint16_t vehicleTypeId)
+    {
+        if (!thingmgr::checkNumFreeThings(12))
```
Comment (duncanspumpkin) [inline src/openloco/things/CreateVehicle.cpp:458]: 3 (vehicle components) * 4 (car components)

File: src/openloco/things/CreateVehicle.cpp [hunk 21]
```diff
@@ -0,0 +1,1000 @@
...
+    static std::optional<uint16_t> sub_4B1E00()
+    {
+        if (!aiIsBelowVehicleLimit())
+        {
+            return {};
+        }
+
+        for (auto i = 0; i < 64 * 1000; i += 64)
+        {
+            auto id = _96885C[i];
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:516]: Add a comment on your suspicions about it having to do with routing

File: src/openloco/things/CreateVehicle.cpp [hunk 22]
```diff
@@ -0,0 +1,1000 @@
...
+        newTail->var_48 = -1;
+        newTail->var_4A = 0;
+        lastVeh->next_car_id = newTail->id;
+        newTail->next_car_id = thing_id::null;
+        return newTail;
+    }
+    // 0x004AE318
+    static std::optional<openloco::vehicle_head*> createBase(const TransportMode mode, const VehicleType type, const uint8_t trackType)
+    {
+        if (!thingmgr::checkNumFreeThings(4))
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:703]: // Head, unk1, unk2, tail

File: src/openloco/things/CreateVehicle.cpp [hunk 23]
```diff
@@ -0,0 +1,1000 @@
...
+
+    static void sub_4AF7A4(openloco::vehicle_head* const veh0)
+    {
+        registers regs{};
+        regs.esi = reinterpret_cast<int32_t>(veh0);
+        call(0x004AF7A4, regs);
+    }
+
+    // 0x00437ED0
+    static void companyRecalculateTransportCounts(const company_id_t companyId)
```
Comment (duncanspumpkin) [inline src/openloco/things/CreateVehicle.cpp:755]: move to company

Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:755]: Move to Company/CompanyMgr

File: src/openloco/things/CreateVehicle.cpp [hunk 24]
```diff
@@ -0,0 +1,1000 @@
...
+    }
+
+    // 0x004B1E77
+    // Free orderId?
+    static void sub_4B1E77(const uint16_t orderId)
+    {
+        uint16_t baseOrderId = orderId & ~(0x3F);
+        for (auto i = 0; i < 64; ++i)
+        {
+            _96885C[i + baseOrderId] = -1;
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:798]: Add constants for this array; -1 and -2

File: src/openloco/things/CreateVehicle.cpp [hunk 25]
```diff
@@ -0,0 +1,1000 @@
...
+        }
+    }
+
+    static void sub_470795(const uint32_t var46, const uint16_t var4C)
+    {
+        auto v = thingmgr::first<openloco::vehicle>();
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:804]: These will all be VehicleHeads

File: src/openloco/things/CreateVehicle.cpp [hunk 26]
```diff
@@ -0,0 +1,1000 @@
...
+            auto next = v->next_vehicle();
+
+            auto head = v->as_vehicle_head();
+            v = next;
+            if (head == nullptr) // Can never happen
+            {
+                continue;
+            }
+
+            if (head->var_46 > var46)
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:816]:
```suggestion
            if (head->var_46 >= var46)
```

Comment (duncanspumpkin) [inline src/openloco/things/CreateVehicle.cpp:816]: `>=`

File: src/openloco/things/CreateVehicle.cpp [hunk 27]
```diff
@@ -0,0 +1,1000 @@
...
+    static void sub_4B1E77(const uint16_t orderId)
+    {
+        uint16_t baseOrderId = orderId & ~(0x3F);
+        for (auto i = 0; i < 64; ++i)
+        {
+            _96885C[i + baseOrderId] = -1;
+        }
+    }
+
+    static void sub_470795(const uint32_t var46, const uint16_t var4C)
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:802]: I think var4C is signed

Comment (duncanspumpkin) [inline src/openloco/things/CreateVehicle.cpp:802]: var4C should be singed.

File: src/openloco/things/CreateVehicle.cpp [hunk 28]
```diff
@@ -0,0 +1,1000 @@
...
+
+    // 0x00470334
+    static void sub_470334(openloco::vehicle_head* const head)
+    {
+        sub_470795(head->var_46, head->var_4C * -1);
+        auto unk = _525FB8 - head->var_46 - head->var_4C;
+        auto var46 = head->var_46;
+        for (; unk != 0; --unk)
+        {
+            _987C5C[var46] = _987C5C[head->var_4C + var46];
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:831]: //TODO: memmove?

File: src/openloco/things/CreateVehicle.cpp [hunk 29]
```diff
@@ -0,0 +1,1000 @@
...
+            var46++;
+        }
+
+        _525FB8 = _525FB8 - head->var_4C;
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:835]: _525FB8 = count of entries in _987C5C?

File: src/openloco/things/CreateVehicle.cpp [hunk 30]
```diff
@@ -0,0 +1,1000 @@
...
+        }
+    }
+
+    // 0x00470334
+    static void sub_470334(openloco::vehicle_head* const head)
+    {
+        sub_470795(head->var_46, head->var_4C * -1);
+        auto unk = _525FB8 - head->var_46 - head->var_4C;
+        auto var46 = head->var_46;
+        for (; unk != 0; --unk)
```
Comment (duncanspumpkin) [inline src/openloco/things/CreateVehicle.cpp:829]: `memmov`

File: src/openloco/things/CreateVehicle.cpp (hunk 25 above)
Comment (duncanspumpkin) [reply src/openloco/things/CreateVehicle.cpp:804]: will be addressed in future refactor

File: src/openloco/things/CreateVehicle.cpp [hunk 31]
```diff
@@ -0,0 +1,966 @@
...
+        regs.al = type;
+        regs.dx = id;
+        call(0x0042851C, regs);
+    }
+
+    // 0x004AE74E
+    static uint32_t create(const uint8_t flags, const uint16_t vehicleTypeId)
+    {
+        gameCommandMapX = location::null;
+        if (!thingmgr::checkNumFreeThings(16))
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:815]: Why 16?

File: src/openloco/things/CreateVehicle.cpp [hunk 32]
```diff
@@ -0,0 +1,966 @@
...
+                auto veh6 = veh2->next_car();
+                thingmgr::freeThing(veh2);
+                thingmgr::freeThing(veh6);
+                return 0x80000000;
+            }
+        }
+        // 0x4AE733
+        auto vehObject = objectmgr::get<vehicle_object>(vehicleTypeId);
+        // TODO: use FixedPoint with 6 {(1 << 6) == 64} decimals for cost_ind
+        auto cost = (vehObject->cost_fact * currencyMultiplicationFactor[vehObject->cost_ind]) / 64;
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:870]: Can we just name these `cost_index` and `cost_factor`?

File: src/openloco/things/CreateVehicle.cpp [hunk 33]
```diff
@@ -0,0 +1,966 @@
...
+            auto _head = *head;
+            _113642A = _head->id;
+            if (createCar(_head, vehicleTypeId))
+            {
+                // 0x004AE6DE
+                sub_4AF7A4(_head);
+                auto company = companymgr::get(_updating_company_id);
+                company->recalculateTransportCounts();
+
+                if (_backupVeh0 != reinterpret_cast<openloco::vehicle_head*>(-1))
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:844]: Is this if ever triggered?

File: src/openloco/things/CreateVehicle.cpp [hunk 34]
```diff
@@ -0,0 +1,966 @@
...
+    static loco_global<uint16_t, 0x009C68E2> gameCommandMapY;
+    static loco_global<uint16_t, 0x009C68E4> gameCommandMapZ;
+    static loco_global<string_id, 0x009C68E6> gGameCommandErrorText;
+    static loco_global<uint8_t, 0x009C68EA> gGameCommandExpenditureType; // premultiplied by 4
+    static loco_global<uint8_t, 0x009C68EE> _errorCompanyId;
+    static loco_global<map::tile_element*, 0x009C68D0> _9C68D0;
+    static loco_global<ColourScheme, 0x01136140> _1136140; // primary colour
+    static loco_global<int32_t, 0x011360FC> _11360FC;
+    static loco_global<openloco::vehicle_head*, 0x01136240> _backupVeh0;
+    static loco_global<int16_t, 0x01136248> _backup2E;
```
Comment (marijnvdwerf) [inline src/openloco/things/CreateVehicle.cpp:43]: Can these backups be locals?

Comment (duncanspumpkin) [reply src/openloco/things/CreateVehicle.cpp:43]: Leave until game command 0 and 6 are implemented.

File: src/openloco/game_commands.h [hunk 35]
```diff
@@ -19,10 +19,20 @@ namespace openloco::game_commands
         flag_6 = 1 << 6, // 0x40
     };
 
+    enum class GameCommand : uint8_t
```
Comment (AaronVanGeffen) [inline src/openloco/game_commands.h:22]: I really like this. While beyond the scope of this PR, we should really add the others as well. (I previously added comments to all game command callers (`do_`).)

Comment (duncanspumpkin) [reply src/openloco/game_commands.h:22]: yeah we were discussing the other game commands and realised we had a bit of knowledge not written down anywhere.

File: src/openloco/things/CreateVehicle.cpp [hunk 36]
```diff
@@ -0,0 +1,966 @@
...
+        _errorCompanyId = company;
+        _9C68D0 = tile == nullptr ? reinterpret_cast<map::tile_element*>(-1) : tile;
+        return false;
+    }
+
+    static bool sub_4B0BDD(openloco::vehicle_head* const head)
+    {
+        switch (head->var_5D)
+        {
+            case 8:
```
[the remark this answers is not in the document]
Comment (duncanspumpkin) [reply src/openloco/things/CreateVehicle.cpp:106]: Will be addressed in a separate PR

**Pull request merged**